### PR TITLE
A better thread-safe profiler constructor and dump profile method

### PR DESCRIPTION
**Pull request by ZihengJiang**
- A better thread-safe singleton `Get` function
- Move `DumpProfile` out of destructor of engine
- Fix operator name bug in bulk execution at graph executor

### CODE REVIEW

File: src/engine/engine.cc [hunk 1]
```diff
@@ -46,6 +47,20 @@ std::shared_ptr<Engine> Engine::_GetSharedRef() {
 
 Engine* Engine::Get() {
   static Engine *inst = _GetSharedRef().get();
+#if MXNET_USE_PROFILER
```
Comment (piiswrong) [inline src/engine/engine.cc]: why put this in engine::get?
consider src/initialize.cc?

File: src/engine/threaded_engine.h [hunk 2]
```diff
@@ -268,13 +268,6 @@ class ThreadedEngine : public Engine {
     objpool_var_ref_    = common::ObjectPool<ThreadedVar>::_GetSharedRef();
   }
   ~ThreadedEngine() {
-#if MXNET_USE_PROFILER
```
Comment (piiswrong) [inline src/engine/threaded_engine.h]: I think naive engine is also using profiler now

Comment (ZihengJiang) [inline src/engine/threaded_engine.h]: I changed NativeEngine also

File: src/initialize.cc [hunk 3]
```diff
@@ -4,13 +4,32 @@
...
 
 namespace mxnet {
 
 class LibraryInitializer {
  public:
   LibraryInitializer() {
     dmlc::InitLogging("mxnet");
+#if MXNET_USE_PROFILER
+    static std::once_flag dump_profile_flag;
+    std::call_once(dump_profile_flag, []() {
```
Comment (piiswrong) [inline src/initialize.cc]: dont need call once since LibraryInitializer will only be called once.

**Comment by piiswrong** [comment]
profiler is not in test suites. Did you test it manually? Could you add a test case?

**Comment by ZihengJiang** [comment]
yep, I've tested it manually. The problem is that I need to change the compile flag to compile with profiler for test case just like MKL. Also, the test case will fail in other build versions without profiler enabled. Do you have any suggestion?

**Comment by piiswrong** [comment]
we can change all test scripts to build with profiler on

**Comment by ZihengJiang** [comment]
um, should I add `-j\$(nproc)` in amalgamation build?

File: Jenkinsfile [hunk 4]
```diff
@@ -95,7 +101,13 @@ USE_CUDNN=1                   \
     node() {
       ws('workspace/amalgamation') {
         init_git()
-        make('cpu', '-C amalgamation/ USE_BLAS=openblas MIN=1')
+        def flag = """ \
+USE_PROFILER=1                \
```
Comment (piiswrong) [inline Jenkinsfile]: don't use profiler for amlagamation

Comment (ZihengJiang) [inline Jenkinsfile]: Will `test_profiler.py` be tested under amlagamation?

File: src/initialize.cc [hunk 5]
```diff
@@ -4,13 +4,28 @@
...
 namespace mxnet {
 
 class LibraryInitializer {
  public:
   LibraryInitializer() {
     dmlc::InitLogging("mxnet");
+#if MXNET_USE_PROFILER
+    // ensure engine's and profiler's constructor are called before atexit.
+    Engine::Get();
+    engine::Profiler::Get();
```
Comment (yajiedesign) [inline src/initialize.cc]: the LibraryInitializer call in dllmain with windows, it will deadlock.please move it to other.

**Pull request merged**